# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/webrender_bindings/DCLayerTree.h
# Commit: ca2e6339f35f
# Full Hash: ca2e6339f35f38cc26a203f29c8dbfeb38f96afc
# Author: Kelsey Gilbert <kelsey.gilbert@mozilla.com>
# Date: 2023-03-14 09:41:39
# Regressor Bug: 1799258
# File Overlap Count: 1
# Description:
#   Bug 1799258 - Do color-management on Windows+DComp via IDCompositionFilterEffects. r=sotaro
#   
#   + Add gfx.color_management.rec709_gamma_as_srgb:true. :'(
#   
#   In particular, rec709(16/255) -> srgb(31/255). Even though it's
# ==============================================================================

diff -r aa4e010199c6 -r ca2e6339f35f gfx/webrender_bindings/DCLayerTree.h
--- a/gfx/webrender_bindings/DCLayerTree.h	Mon Mar 13 21:04:10 2023 +0000
+++ b/gfx/webrender_bindings/DCLayerTree.h	Mon Mar 13 21:04:10 2023 +0000
@@ -12,6 +12,7 @@
 #include <vector>
 #include <windows.h>
 
+#include "Colorspaces.h"
 #include "GLTypes.h"
 #include "mozilla/HashFunctions.h"
 #include "mozilla/layers/OverlayInfo.h"
@@ -27,7 +28,11 @@
 struct ID3D11VideoProcessor;
 struct ID3D11VideoProcessorEnumerator;
 struct ID3D11VideoProcessorOutputView;
+struct IDCompositionColorMatrixEffect;
+struct IDCompositionFilterEffect;
+struct IDCompositionTableTransferEffect;
 struct IDCompositionDevice2;
+struct IDCompositionDevice3;
 struct IDCompositionSurface;
 struct IDCompositionTarget;
 struct IDCompositionVisual2;
@@ -66,6 +71,23 @@
   UINT mRgb10a2OverlaySupportFlags = 0;
 };
 
+// -
+
+struct ColorManagementChain {
+  RefPtr<IDCompositionColorMatrixEffect> srcRgbFromSrcYuv;
+  RefPtr<IDCompositionTableTransferEffect> srcLinearFromSrcTf;
+  RefPtr<IDCompositionColorMatrixEffect> dstLinearFromSrcLinear;
+  RefPtr<IDCompositionTableTransferEffect> dstTfFromDstLinear;
+  RefPtr<IDCompositionFilterEffect> last;
+
+  static ColorManagementChain From(IDCompositionDevice3& dcomp,
+                                   const color::ColorProfileConversionDesc&);
+
+  ~ColorManagementChain();
+};
+
+// -
+
 /**
  * DCLayerTree manages direct composition layers.
  * It does not manage gecko's layers::Layer.
@@ -210,6 +232,19 @@
 
   bool mPendingCommit;
 
+  static color::ColorProfileDesc QueryOutputColorProfile();
+
+  mutable Maybe<color::ColorProfileDesc> mOutputColorProfile;
+
+ public:
+  const color::ColorProfileDesc& OutputColorProfile() const {
+    if (!mOutputColorProfile) {
+      mOutputColorProfile = Some(QueryOutputColorProfile());
+    }
+    return *mOutputColorProfile;
+  }
+
+ protected:
   static UniquePtr<GpuOverlayInfo> sGpuOverlayInfo;
 };
 
@@ -322,6 +357,7 @@
 
   UniquePtr<DCSurface> mSurface;
   const bool mIsOpaque;
+  Maybe<ColorManagementChain> mCManageChain;
 };
 
 class DCSurfaceVideo : public DCSurface {